internal/lock.go

```diff
@@ -26,178 +28,295 @@ type config struct {
 	Resource []Resource
 }
 
+// NewLock creates a new lock file instance
 func NewLock(path string, newOk bool) (*Lock, error) {
 	_, error := os.Stat(path)
 	if os.IsNotExist(error) {
 		if newOk {
-			return &Lock{path: path}, nil
+			return &Lock{path: path, conf: config{Resource: []Resource{}}}, nil
 		} else {
 			return nil, fmt.Errorf("file '%s' does not exist", path)
 		}
 	}
+
 	var conf config
 	file, err := os.Open(path)
 	if err != nil {
-		return nil, error
+		return nil, err
 	}
+	defer file.Close()
+
 	d := toml.NewDecoder(file)
-	err = d.Decode(&conf)
-	if err != nil {
-		return nil, err
+	if err := d.Decode(&conf); err != nil {
+		return nil, fmt.Errorf("failed to decode lock file: %w", err)
 	}
 
 	return &Lock{path: path, conf: conf}, nil
 }
 
+// AddResource adds a resource without cache
 func (l *Lock) AddResource(paths []string, algo string, tags []string, filename string) error {
+	return l.AddResourceWithCache(paths, algo, tags, filename, "")
+}
+
+// AddResourceWithCache adds a resource with cache
+func (l *Lock) AddResourceWithCache(paths []string, algo string, tags []string, filename string, cacheUri string) error {
 	for _, u := range paths {
 		if l.Contains(u) {
 			return fmt.Errorf("resource '%s' is already present", u)
 		}
 	}
-	r, err := NewResourceFromUrl(paths, algo, tags, filename)
+
+	log.Debug().
+		Strs("paths", paths).
+		Str("algo", algo).
+		Strs("tags", tags).
+		Str("filename", filename).
+		Str("cache_uri", cacheUri).
+		Msg("Adding new resource")
+
+	r, err := NewResourceFromUrlWithCache(paths, algo, tags, filename, cacheUri)
 	if err != nil {
-		return err
+		log.Error().Err(err).Msg("Failed to create resource")
+		return fmt.Errorf("failed to create resource: %w", err)
```

Comment:
I'll stop making this same comment after this one: is the error message returned from the failing function insufficient?  If so, should we fix the error message there?